[BUG] Support custom LLVM installation locations

My Fedora system's LLVM is too old (10). And I have to compile my own LLVM from source (version 14) and install under `/opt/llvm/`. Currently, the LLVM JIT build of bpftime only looks for system llvm locations, which won't work for my situation:

```
CMake Error at vm/llvm-jit/CMakeLists.txt:46 (find_package):
  Could not find a package configuration file provided by "LLVM" with any of
  the following names:

    LLVMConfig.cmake
    llvm-config.cmake

  Add the installation prefix of "LLVM" to CMAKE_PREFIX_PATH or set
  "LLVM_DIR" to a directory containing one of the above files.  If "LLVM"
  provides a separate development package or SDK, be sure it has been
  installed.
```

Hopefully there is a way to configure the LLVM_DIR without editing any files.

--- a/.github/workflows/test-aot-cli.yml
+++ b/.github/workflows/test-aot-cli.yml
@@ -32,10 +32,14 @@ jobs:
         submodules: 'recursive'
     - name: Build and install everything (Ubuntu)
       if: startsWith(matrix.runs-on-container,'ubuntu')
+      env:
+        LLVM_ROOT: /usr/lib/llvm-15
       run: |
         CC=gcc-12 CXX=g++-12 make release-with-llvm-jit -j
     - name: Build and install everything (Fedora)
       if: startsWith(matrix.runs-on-container,'fedora')
+      env:
+        LLVM_ROOT: /usr/lib64/llvm15
       run: |
         make release-with-llvm-jit -j
     - name: Do compilation & run

--- a/.github/workflows/test-examples.yml
+++ b/.github/workflows/test-examples.yml
@@ -37,8 +37,16 @@ jobs:
       - uses: actions/checkout@v2
         with:
           submodules: 'recursive'
-      - name: Build and install runtime (with llvm-jit)
-        if: ${{matrix.enable_jit}}
+      - name: Build and install runtime (with llvm-jit and ubuntu)
+        if: ${{matrix.enable_jit && startsWith(matrix.runs-on-container.name,'ubuntu')}}
+        env:
+          LLVM_ROOT: /usr/lib/llvm-15
+        run: |
+          ${{matrix.runs-on-container.compiler}} make release-with-llvm-jit -j
+      - name: Build and install runtime (with llvm-jit and fedora)
+        if: ${{matrix.enable_jit && startsWith(matrix.runs-on-container.name,'fedora')}}
+        env:
+          LLVM_ROOT: /usr/lib64/llvm15
         run: |
           ${{matrix.runs-on-container.compiler}} make release-with-llvm-jit -j
       - name: Build and install runtime (without llvm-jit)

--- a/.github/workflows/test-llvm-jit.yml
+++ b/.github/workflows/test-llvm-jit.yml
@@ -32,10 +32,14 @@ jobs:
         submodules: 'recursive'
     - name: Build and install everything (Ubuntu)
       if: startsWith(matrix.runs-on-container,'ubuntu')
+      env:
+        LLVM_ROOT: /usr/lib/llvm-15
       run: |
         CC=gcc-12 CXX=g++-12 make build-llvm -j
     - name: Build and install everything (Fedora)
       if: startsWith(matrix.runs-on-container,'fedora')
+      env:
+        LLVM_ROOT: /usr/lib64/llvm15
       run: |
         make build-llvm -j
     - name: Run tests
@@ -44,6 +48,8 @@ jobs:
 
     - name: build llvm JIT/AOT as a standalone library (Ubuntu)
       if: startsWith(matrix.runs-on-container,'ubuntu')
+      env:
+        LLVM_ROOT: /usr/lib/llvm-15
       run: |
         cd vm/llvm-jit &&\
         CC=gcc-12 CXX=g++-12 cmake -B build -DCMAKE_BUILD_TYPE=Release &&\
@@ -51,10 +57,14 @@ jobs:
 
     - name: build vm as a standalone library (Ubuntu)
       if: startsWith(matrix.runs-on-container,'ubuntu')
+      env:
+        LLVM_ROOT: /usr/lib/llvm-15
       run: |
         cd vm && CC=gcc-12 CXX=g++-12 make build-llvm -j
     - name: build llvm JIT/AOT as a standalone library (Fedora)
       if: startsWith(matrix.runs-on-container,'fedora')
+      env:
+        LLVM_ROOT: /usr/lib64/llvm15
       run: |
         cd vm/llvm-jit &&\
         cmake -B build -DCMAKE_BUILD_TYPE=Release &&\
@@ -62,6 +72,8 @@ jobs:
 
     - name: build vm as a standalone library (Fedora)
       if: startsWith(matrix.runs-on-container,'fedora')
+      env:
+        LLVM_ROOT: /usr/lib64/llvm15
       run: |
         cd vm && make build-llvm -j
     

--- a/.github/workflows/test-vm.yml
+++ b/.github/workflows/test-vm.yml
@@ -43,14 +43,37 @@ jobs:
     - uses: actions/checkout@v2
       with:
         submodules: 'recursive'
-    - name: build
+    - name: build (Ubuntu)
+      if: startsWith(matrix.runs-on-container,'ubuntu')
+      env:
+        LLVM_ROOT: /usr/lib/llvm-15
+      run: make build-vm
+    - name: build (Fedora)
+      if: startsWith(matrix.runs-on-container,'fedora')
+      env:
+        LLVM_ROOT: /usr/lib64/llvm15
       run: make build-vm
 
-    - name: run tests x86
+    - name: run tests x86 (Ubuntu)
       shell: bash
+      if: startsWith(matrix.runs-on-container,'ubuntu')
+      env:
+        LLVM_ROOT: /usr/lib/llvm-15
       run: |
         python3.8 -m venv vm/test
         source vm/test/bin/activate
         pip install -r vm/test/requirements.txt
         # make build # or build-arm32 build-arm64
         make -C vm test-vm -j
+    - name: run tests x86 (Fedora)
+      shell: bash
+      if: startsWith(matrix.runs-on-container,'fedora')
+      env:
+        LLVM_ROOT: /usr/lib64/llvm15
+      run: |
+        python3.8 -m venv vm/test
+        source vm/test/bin/activate
+        pip install -r vm/test/requirements.txt
+        # make build # or build-arm32 build-arm64
+        make -C vm test-vm -j
+

--- a/vm/llvm-jit/CMakeLists.txt
+++ b/vm/llvm-jit/CMakeLists.txt
@@ -37,23 +37,22 @@ bpftime_add_library(vm-bpf
 
 set_target_properties(vm-bpf PROPERTIES ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/../")
 
-set(LLVM_SEARCH_PATHS
-    /usr/lib/llvm-15/lib/cmake/llvm
-    /usr/lib/llvm-14/lib/cmake/llvm
-    /usr/lib/llvm-13/lib/cmake/llvm
-    /usr/lib64/llvm15/lib/cmake/llvm
-)
+find_package(LLVM REQUIRED CONFIG)
 
-find_package(LLVM REQUIRED CONFIG
-    PATHS ${LLVM_SEARCH_PATHS}
-    NO_DEFAULT_PATH)
+if (${LLVM_PACKAGE_VERSION} VERSION_LESS 15)
+    message(FATAL_ERROR "LLVM version must be >=15")
+endif()
 
-llvm_map_components_to_libnames(LLVM_LIBS
-    Core
-    OrcJIT
-    Support
-    nativecodegen
-)
+if(ENABLE_LLVM_SHARED)
+    set(LLVM_LIBS LLVM)
+else()
+    llvm_map_components_to_libnames(LLVM_LIBS
+        Core
+        OrcJIT
+        Support
+        nativecodegen
+    )
+endif()
 
 if(NOT DEFINED SPDLOG_INCLUDE)
   message(INFO "Adding spdlo seperately..")

